Validator [MAILPOET-4195]

JanJakes opened the pull request:
[MAILPOET-4195]

[MAILPOET-4195]: https://mailpoet.atlassian.net/browse/MAILPOET-4195?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

For details see https://wp.me/pcNwfB-os.

Usage:
```php
$schema = Builder::object()
  ->title('User')
  ->description('User schema definition')
  ->field('name', 'user')
  ->field('version', 1)
  ->properties([
    'id' => Builder::string()->required()->formatUuid(),
    'created_at' => Builder::string()->required()->formatDateTime(),
    'username' => Builder::string()->required()->minLength(2)->maxLength(30)->pattern('^[a-z0-9]+$'),
    'password' => Builder::string()->required()->minLength(8)->maxLength(1024),
    'email' => Builder::string()->required()->formatEmail(),
    'ip' => Builder::string()->required()->formatIp(),
    'refresh_interval' => Builder::integer()->required()->multipleOf(3600),
    'subscribed' => Builder::boolean()->default(false),

    // nested object
    'profile' => Builder::object([
      'url' => Builder::string()->required()->formatUri(),
      'photo_url' => Builder::string()->required()->nullable(),
      'color' => Builder::string()->required()->formatHexColor(),
      'age' => Builder::integer()->required(),
      'rating' => Builder::integer()->required()->minimum(0)->maximum(5),
      'score' => Builder::number()->required()->exclusiveMinimum(0)->maximum(100),
      'distance' => Builder::number()->required(),
    ]),

    // array of unique objects
    'preferences' => Builder::array(
      Builder::object([
        'key' => Builder::string()->required(),
        'value' => Builder::anyOf([Builder::string(), Builder::integer(), Builder::number()])->required(),
        'meta' => Builder::string(),
      ])
    )->uniqueItems(),

    // pattern properties
    'properties' => Builder::object()->patternProperties([
      '^number_' => Builder::anyOf([Builder::integer(), Builder::number()]),
      '^string_' => Builder::string(),
      '^bool_' => Builder::boolean(),
    ]),

    // oneOf
    'linked_accounts' => Builder::array(
      Builder::oneOf([
        Builder::object([
          'apple_id' => Builder::string()->required(),
        ]),
        Builder::object([
          'facebook_id' => Builder::string()->required(),
        ]),
        Builder::object([
          'google_id' => Builder::string()->required(),
        ]),
      ])
    ),

    // additional properties, anyOf
    'attributes' => Builder::object()->additionalProperties(
      Builder::anyOf([
        Builder::string(),
        Builder::integer(),
        Builder::boolean(),
        Builder::number(),
        Builder::null(),
      ])
    ),
]);

$value = [
  'id' => '4c30b777-9067-4a7b-b498-e295eabbf6ef',
  'created_at' => '2022-03-23T15:21:37',
  'username' => 'mailpoet',
  'password' => '4f38f3eb301923fd38f3',
  'email' => 'mailpoet@example.com',
  'ip' => '145.8.23.191',
  'refresh_interval' => 7200,
  'subscribed' => false,
  'profile' => [
    'url' => 'https://example.com',
    'photo_url' => null,
    'color' => '#a5a5a5',
    'age' => 38,
    'rating' => 4,
    'score' => 87.372,
    'distance' => 1023.28,
  ],
  'preferences' => [
    ['key' => 'theme', 'value' => 'dark'],
    ['key' => 'screen_with', 'value' => 1170, 'meta' => 'ios/safari'],
    ['key' => 'screen_with', 'value' => 3072, 'meta' => 'macos/chrome'],
  ],
  'properties' => [
    'bool_tracking_enabled' => true,
    'number_tracking_page_views' => 12,
    'string_tracking_id' => 'az0x8kw25as',
  ],
  'linked_accounts' => [
    ['apple_id' => 'f794613bd965'],
    ['google_id' => 'f794613bd965'],
  ],
  'attributes' => [
    'browser' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36',
    'language' => 'en_US',
    'timezone' => 'Europe/Prague',
    'login_attempts' => 2,
    'is_admin' => false,
    'referrer' => null,
    'ping' => 173,
    'risk_score' => 0.10378,
  ],
];

try {
  $data = $validator->validate($schema, $value);
} catch (ValidationException $e) {
  // process errors
  $e->getWpError();
}

// use sanitized $data

// JSON schema as plain array (WP-compatible)
$schema->toArray();

// JSON schema as JSON string (both options work)
$schema->toString();
(string)$schema;
```

johnolek commented on mailpoet/lib/Validator/Validator.php:
```diff
@@ -0,0 +1,190 @@
+    if ($result instanceof WP_Error) {
+      throw ValidationException::createFromWpError($result);
+    }
+    return $result;
```
This is a really basic and maybe silly question that probably demonstrates that I know nothing about what we're trying to achieve here, but why do we want to have validation and sanitization combined like this? Why not have 

```php
/**
 * @param mixed $value
 * @throws ValidationException
 */
public function validate(Schema $schema, $value, string $paramName = 'value'): void {
  $result = $this->validateAndSanitizeValueFromSchema($value, $schema->toArray(), $paramName);
  if ($result instanceof WP_Error) {
    throw ValidationException::createFromWpError($result);
  }
}
```

and then continue to use the original value if validation doesn't throw an exception?

It seems like if it passes validation it should be fine to use as-is, and then we don't have to worry about the order of schemas in `anyOf` definitions.

I suppose we always have the option to ignore the return value and continue using the original if we feel like it. I just want to make sure I understand why it's assumed we would always use the sanitized version.

JanJakes replied:
Using the code snippet you've shared it would be impossible to get the sanitized version. In the current form, you're free to choose whether you use the sanitized version or the original one.

But given the discussed issues with integers and floats and the fact that WP performs further sanitization (e.g. a string of type URL gets normalized to URL-encoded version), it's always strongly recommended to use the sanitized data. As we can [find in the WP docs](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#api):

> When calling these functions, you should take care to always first validate the data using rest_validate_value_from_schema, and then if that function returns true, sanitize the data using rest_sanitize_value_from_schema. Not using both can open up your endpoint to security vulnerabilities.

johnolek commented on mailpoet/tests/integration/Validator/ValidatorTest.php:
```diff
@@ -0,0 +1,745 @@
+    $this->assertValidationFailed(Builder::number(), [], 'value is not of type number.');
+    $this->assertValidationFailed(Builder::number(), new stdClass(), 'value is not of type number.');
+
+    $this->assertValidationFailed(Builder::array(), 'a,b', 'value is not of type array.');
+    $this->assertValidationFailed(Builder::array(), ['x' => 'x', 'y' => 'y'], 'value is not of type array.');
+    $this->assertValidationFailed(Builder::object(), '', 'value is not of type object.');
```
Are these in the correct test?

JanJakes replied:
Fixed.

JanJakes commented on mailpoet/lib/Validator/Schema.php:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  /** @return static */
+  public function title(?string $title) {
+    return $title === null
+      ? $this->unsetSchemaProperty('title')
+      : $this->updateSchemaProperty('title', $title);
+  }
+
+  /** @return static */
+  public function description(?string $description) {
+    return $description === null
+      ? $this->unsetSchemaProperty('description')
+      : $this->updateSchemaProperty('description', $description);
+  }
```
@johnolek Good question, I don't know 😆 I wanted to keep it "reversible" but maybe it's not needed. The general idea is that you can have a reusable schema in a variable but maybe want to modify it somehow. But yeah, no real use-case at the moment. Should we keep it/remove it? Should I add tests?

JanJakes replied:
Added tests for all the methods that use `unsetSchemaProperty` in this class.

JanJakes commented on do:
```diff
@@ -90,5 +90,5 @@ else
     params=("${params[@]:1}")
   fi
-  ssh_and_run $docker_service "cd wp-content/plugins/$plugin_directory && ./do ${params[@]}"
+  ssh_and_run $docker_service "cd wp-content/plugins/$plugin_directory && ./do ${params[*]}"
```
Definitely not. I think we don't have any clear rules whether crating new tickets and PRs is always necessary for small fixes. We've sometimes added small improvements in the past, as I did here, considering they're in a separate commit that makes it clear what it does.

In this case, for instance, it's a tiny fix (and similarly for the exception return types) and probably an obvious bug, and creating a ticket for each... I'm not sure, maybe it's overkill?

JanJakes commented on mailpoet/lib/Validator/Schema.php:
```diff
@@ -0,0 +1,102 @@
+    return $json;
+  }
+
+  public function __toString(): string {
```
It was just an initial idea in the API but later I added `toString()` to keep it unified with `toArray()`. I'm OK with removing it, I guess 👍

JanJakes commented on mailpoet/lib/Validator/Schema.php:
```diff
@@ -0,0 +1,102 @@
+      : $this->updateSchemaProperty($name, $value);
+  }
+
+  public function toArray(): array {
+    return $this->schema;
+  }
```
You're right, it doesn't seem to be used though. So I don't know... should we add it somehow? The thing is we can't make it being added automatically because the schema factories don't know which one of them is root 🤔

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+  ];
+
+  /** @param array<string, Schema> $properties */
+  public function properties(array $properties): self {
```
We can do that, I was probably thinking that since there are also pattern and additional properties, I'd just leave a method for that. I was also thinking about different APIs for making required/optional properties easier to write.

JanJakes replied:
Looking at it... if I add the constructor, I should probably add one to the array schema as well. Maybe it's fine as it is, considering `Builder` will be used anyway? The difference from `oneOf`/`anyOf` is they don't really make much sense without subschemas while types like object and array can be used without property and item definitions.

johnolek replied:
Very good point about empty objects/arrays being valid! And `oneOf`/`anyOf` do feel like special cases.

On that note, you are allowing `oneOf`/`anyOf` to be instantiated with an empty array of schemas while the `Builder` doesn't allow for that. Should those things be made consistent? Probably not important if the Builder is the main interface for using those schemas, but it does seem a little strange to allow schemas that don't make much sense.

JanJakes replied:
Good point, I don't know why I had the default values for `oneOf`/`anyOf` in the first place. I removed the defaults.

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  /** @param array<string, Schema> $properties */
+  public function patternProperties(array $properties): self {
```
That's a good idea. Strictly following JSON schema specs we should validate for ECMA 262 regex specs but I'm not sure if there's an easy way to do that (e.g. a regex that validates ECMA 262 regex), can't find any with a quick search.

But maybe... since we're in PHP and WP uses PHP regexes internally anyway, we could simply try to match it against an empty string and catch any errors. I think that would be a good addition 👍

JanJakes replied:
Added a new commit to address this for both string patterns and pattern properties.

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  /** @param array<string, Schema> $properties */
+  public function patternProperties(array $properties): self {
```
I don't know how that happens but GitHub shows me this kind of duplicity with https://github.com/mailpoet/mailpoet/pull/4027#discussion_r834642243 so I answered there.

JanJakes replied:
Added a new commit to address this for both string patterns and pattern properties.

johnolek replied:
I worried that might happen. I thought I submitted the first comment, but then accidentally navigated away from the page. When I came back the comment wasn't showing up so I rewrote it, and apparently both actually got saved.

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+    return $this->updateSchemaProperty('additionalProperties', false);
+  }
+
+  /** @param array<string, Schema> $properties */
```
That's a good point, I'll have a look and add more links.

> Some of the sections are in confusing places, like [format](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#format) only seems to apply to strings yet it's not in the Strings section.
This may originate from a bit of a strange nature of JSON schema itself. It's completely valid to specify `type` as `["string", "integer", "number", "boolean", ...]` and then add things like `format` or `minimum`, etc. Each of those modifiers will be applied only on the type that supports it though and for `format` it's indeed just for strings.

The schema builder I created is more restrictive and doesn't allow this type of mixing so it's clear what's being used and what's possible for each type. It shouldn't compromise on flexibility as we still have `oneOf`/`anyOf`, etc.

JanJakes replied:
Added info for object `patternProperties` as well as string `pattern`.

johnolek replied:
Thanks for the clarification on the JSON schema and for adding extra info!

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+    return $this->updateSchemaProperty('additionalProperties', $schema->toArray());
+  }
+
+  public function enableAdditionalProperties(): self {
+    if (($this->schema['additionalProperties'] ?? []) === false) {
+      return $this->unsetSchemaProperty('additionalProperties');
+    }
+    return clone $this;
+  }
```
Using `true` would work as well but I was thinking more like:
```
$schema1->toString() === $schema1->enableAdditionalProperties()->toString()
```
So I simply wanted to make a "consistent" output for semantically identical schemas.

johnolek replied:
That makes total sense and sounds good to me!

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public function enableAdditionalProperties(): self {
+    if (($this->schema['additionalProperties'] ?? []) === false) {
```
Additional properties can either have a schema or boolean (or just `false`? now I'm not sure). So considering you've already set `additionalProperties` to some schema, calling `enableAdditionalProperties` after it shouldn't unset it, I guess.

JanJakes replied:
@johnolek Ah, right! That's a very good point. I renamed the method to `enableAnyAdditionalProperties` and made it simply unset the schema field without any complex logic or checking whether something is already set.

JanJakes replied:
@johnolek So one more update — based on Rosťa's comments ([here](https://github.com/mailpoet/mailpoet/pull/4027#discussion_r836173225) and [here](https://github.com/mailpoet/mailpoet/pull/4027#discussion_r836181569)), I've decided to remove all the "unsetters" (except for "optional" and "nonNullable") for now. We can always add them if the need arises.

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,55 @@
+use MailPoet\Validator\Schema;
+
+// See: https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#objects
+class ObjectSchema extends Schema {
```
Not sure, maybe we can decide on this later? So far it's all immutable so I guess the API would be more like `withProperty` or `withoutProperty` to keep the immutability 🤔

johnolek replied:
This is a good example of me coming up with an idea that there's no actual need for at the moment. Let's definitely keep this as simple as we can for now and only consider extra methods like I mentioned if it ever comes up (which it almost certainly won't).

JanJakes commented on mailpoet/lib/Validator/Schema/OneOfSchema.php:
```diff
@@ -0,0 +1,41 @@
+  public function nonNullable(): self {
+    $null = ['type' => 'null'];
+    $oneOf = $this->schema['oneOf'];
+    $value = in_array($null, $oneOf, true)
```
I think you're right 👍 Will fix.

JanJakes replied:
Fixed.

JanJakes commented on mailpoet/lib/Validator/Validator.php:
```diff
@@ -0,0 +1,190 @@
+
+  /**
+   * Strict validation & sanitization implementation.
+   * It only coerces int to float (e.g. 5 to 5.0).
```
Good question. I think the main reason is not coercion but validation. When adding the strict type checks, I couldn't simply do `is_float` but accept integers as well. Then I left the coercion to the internal WP sanitizer and it indeed normalizes it to floats.

But I think that's correct — we may not want to get sometimes integer and sometimes float for the same property, based only on whether the decimal part is zero or not. I was also using [this handbook](https://json-schema.org/understanding-json-schema/reference/numeric.html#number) a lot and it seems in both Python and Ruby they handle it as floats.

So yeah, the coercion part may not be 100% necessary but I guess it's safer and ensures returning just one type.

johnolek replied:
Makes sense to me. Consistent types are very nice, and it's good to know other languages handle it the same way.

JanJakes commented on mailpoet/lib/Validator/Validator.php:
```diff
@@ -0,0 +1,190 @@
+   * @param mixed $value
+   * @return mixed|WP_Error
+   */
+  private function findAnyMatchingSchema($value, array $anyOfSchema, string $paramName) {
+    $errors = [];
+    foreach ($anyOfSchema['anyOf'] as $index => $schema) {
+      $result = $this->validateAndSanitizeValueFromSchema($value, $schema, $paramName);
+      if (!$this->wp->isWpError($result)) {
+        return $result;
+      }
+      $errors[] = ['error_object' => $result, 'schema' => $schema, 'index' => $index];
+    }
+    return $this->wp->restGetCombiningOperationError($value, $paramName, $errors);
+  }
+
+  /**
+   * @param mixed $value
+   * @return mixed|WP_Error
+   */
+  private function findOneMatchingSchema($value, array $oneOfSchema, string $paramName) {
```
Good point, originally I copied them from `rest-api.php` but amended them to carry the sanitized values so yep, now I should maybe rename these. Will have a look.

JanJakes replied:
Renamed.

JanJakes commented on mailpoet/lib/Validator/Validator.php:
```diff
@@ -0,0 +1,190 @@
+
+  /** @param string|string[] $type */
+  private function getTypeError(string $param, $type): WP_Error {
+    $type = is_array($type) ? $type : [$type];
```
Not sure, I had to try. In this case, it seems there is not: https://3v4l.org/v3MXU

But I wonder what's better to keep for readability...

johnolek replied:
Good point. I like it as-is because it's more explicit.

JanJakes commented on mailpoet/tests/integration/Validator/ValidatorTest.php:
```diff
@@ -0,0 +1,745 @@
+
+  public function testNumber(): void {
+    // valid
+    $this->assertValidationPassed(Builder::number(), 5, 5.0); // 5.0
+    $this->assertValidationPassed(Builder::number(), 0.123);
+    $this->assertValidationPassed(Builder::number(), 1e3);
+    $this->assertValidationPassed(Builder::number(), -5, -5.0); // -5.0
+    $this->assertValidationPassed(Builder::number(), -0.123);
+    $this->assertValidationPassed(Builder::number(), -1e3);
+    $this->assertValidationPassed(Builder::number(), 0, 0.0); // 0.0
+    $this->assertValidationPassed(Builder::number(), -0, 0.0); // 0.0
```
Good catch, those were there for me before it was behaving as expected.

JanJakes replied:
Fixed.

JanJakes commented:
@johnolek Thanks for such a thorough review and great suggestions. I addressed all issues I considered immediately actionable and repushed the branch. Most of them were small so I amended respective commits. I only added a new commit for regex validation.

costasovo commented on mailpoet/lib/Validator/Validator.php:
```diff
@@ -0,0 +1,190 @@
+        foreach ($value as $k => $v) {
+          if (isset($schema['properties'][$k])) {
+            $result = $this->validateAndSanitizeValueFromSchema($v, $schema['properties'][$k], $paramName . '[' . $k . ']');
+            if ($this->wp->isWpError($result)) {
+              return $result;
+            }
```
If I'm not mistaken we end validation with the first error here. What about collecting errors from all properties and combining  them using `$this->wp->restGetCombiningOperationError($value, $paramName, $errors);` like we do in `validateAndSanitizeAnyOf`? I think it might be handy to get all errors at once.

JanJakes replied:
This is only mirroring the behavior of WordPress. I'll add some comments on which functions are mirrored.

costasovo replied:
Ok. Let's keep the behavior similar to what WP does.

costasovo reviewed (commented):
Cool stuff! 👏 

I added just a few questions.

JanJakes commented on mailpoet/lib/Validator/Schema/ArraySchema.php:
```diff
@@ -0,0 +1,32 @@
+    return $this->updateSchemaProperty('uniqueItems', true);
+  }
+
+  public function nonUniqueItems(): self {
```
👍  Let's remove those for now, together with other "unsetters". We can always add such methods later on.

JanJakes commented on mailpoet/lib/Validator/Schema/ObjectSchema.php:
```diff
@@ -0,0 +1,57 @@
+    return $this->updateSchemaProperty('additionalProperties', $schema->toArray());
+  }
+
+  public function enableAnyAdditionalProperties(): self {
+    return $this->unsetSchemaProperty('additionalProperties');
+  }
```
👍  Let's remove those for now, together with other "unsetters". We can always add such methods later on.

costasovo reviewed (approved):
👍
